[docs-only] add more general API docs (#5301)
owncloud/ocis: docs/apis/_index.md

geekdocCollapseSection: true
---

{{< toc >}}

Infinite Scale provides a large set of different **application programming interfaces (APIs)**. Infinite Scale is built by microservices. That means many calls to "functions" in the code are remote calls.

Basically we have two different API "universes": [HTTP](http) and [gRPC](grpc_apis).

{{< columns >}} <!-- begin columns block -->

{{< figure src="/ocis/static/http-logo.png" width="70%" >}}
<--->

{{< figure src="/ocis/static/grpc-logo.png" width="70%" >}}

{{< /columns >}}

> All our clients talk to the Server using Http Apis. This has legacy reasons and is well-supported across many platforms and technologies. Infinite Scale uses an [HTTP API gateway](../services/proxy) to route client requests to the correct service. For inter-service-communication we are using mostly gRPC calls because it has some advantages. In the future, it may occur that clients use gRPC directly to gain some advantages.

Author: thanks. Corrected.

### RFC

Some APIs have become a de facto standard and are additionally covered by an [RFC](https://en.wikipedia.org/wiki/Request_for_Comments).

## [gRPC](grpc_apis)

In gRPC, a client application can directly call methods on a server application on a different machine as if it was a local object. This makes it easier to create distributed applications based on microservices. In gRPC we can define a service and specify the methods that can be called remotely. A gRPC client has a stub that provides the same methods and types as the server.
Infinite Scale uses an [gRPC API Gateway](../services/gateway) to route the requests to the correct service.

### Protobuf

gRPC APIs are typically defined by [Protocol buffers](https://developers.google.com/protocol-buffers/docs/overview). The different client and server stubs are created from ``*.proto`` files by code generation tools.

## Versioning

> There are different standards for API versioning: Through URL, trough request parameter, through custom header and through content negotiation. Ocis uses the versioning by URL scheme although this strategy creates a big code footprint. The versioning follows [SemVer](https://semver.org). We update the major version number when breaking changes are needed. Clients can decide which major version they use through the request URL. The specific implementation is documented on each API.

Reviewer: ```suggestion
There are different standards for API versioning: Through URL, through request parameter, through custom header and through content negotiation. Ocis uses the versioning by URL scheme although this strategy creates a big code footprint. The versioning follows [SemVer](https://semver.org). We update the major version number when breaking changes are needed. Clients can decide which major version they use through the request URL. The specific implementation is documented on each API.
```

geekdocCollapseSection: true
---

{{< toc-tree >}}

Infinite Scale provides a large set of different **application programming interfaces (APIs)**. Infinite Scale is built by microservices. That means many calls to "functions" in the code are remote calls.

Basically we have two different API "universes": [HTTP](http) and [gRPC](grpc_apis).

{{< columns >}} <!-- begin columns block -->

{{< figure src="/ocis/static/http-logo.png" width="70%" alt="Image sourced from https://commons.wikimedia.org/ free of licenses" >}}
<--->

{{< figure src="/ocis/static/grpc-logo.png" width="70%" alt="Image sourced from https://grpc.io/ under CC 4.0 BY license" >}}

{{< /columns >}}

> For inter-service-communication we are using mostly gRPC calls because it has some advantages. In the future, it may occur that clients use gRPC directly to make use of these advantages.

Reviewer: ```suggestion
For inter-service-communication we are using mostly gRPC calls because it has some advantages. In the future, clients may decide to use gRPC directly to make use of these advantages.
```

It is best practise to define APIs and their behavior by a spec. We are using the OpenAPI standard for all new APIs. The [OpenAPI Specification](https://swagger.io/specification/), previously known as the Swagger Specification, is a specification for a machine-readable interface definition language for describing, producing, consuming and visualizing RESTful web services. Previously part of the Swagger framework, it became a separate project in 2016, overseen by the OpenAPI Initiative, an open-source collaboration project of the Linux Foundation. Swagger and some other tools can generate code, documentation and test cases from interface files.

### RFC

Some APIs have become a de facto standard and are additionally covered by an [RFC](https://en.wikipedia.org/wiki/Request_for_Comments).

## [gRPC](grpc_apis)

In gRPC, a client application can directly call methods on a server application on a different machine as if it was a local object. This makes it easier to create distributed applications based on microservices. In gRPC we can define a service and specify the methods that can be called remotely. A gRPC client has a stub that provides the same methods and types as the server.

> Infinite Scale uses an [gRPC API Gateway](../services/gateway) to route the requests to the correct service.

Reviewer: ```suggestion
Infinite Scale uses a [gRPC API Gateway](../services/gateway) to route the requests to the correct service.
```
